mavros_extras/src/plugins/adsb.cpp

```diff
@@ -0,0 +1,121 @@
+/**
+ * @brief Automatic dependent surveillance-broadcast Vehicle plugin
+ * @file adsb.cpp
+ * @author Nuno Marques <n.marques21@hotmail.com>
+ *
+ * @addtogroup plugin
+ * @{
+ */
+/*
+ * Copyright 2017 Nuno Marques.
+ *
+ * This file is part of the mavros package and subject to the license terms
+ * in the top-level LICENSE file of the mavros repository.
+ * https://github.com/mavlink/mavros/tree/master/LICENSE.md
+ */
+
+#include <mavros/mavros_plugin.h>
+
+#include <mavros_msgs/ADSBVehicle.h>
+
+namespace mavros {
+namespace extra_plugins {
+
+/**
+ * @brief ADS-B Vehicle plugin
+ *
+ * Publish/subscribe Automatic dependent surveillance-broadcast data to/from a vehicle.
+ */
+class ADSBPlugin : public plugin::PluginBase {
+public:
+	ADSBPlugin() : PluginBase(),
+		adsb_nh("~adsb")
+	{ }
+
+	void initialize(UAS &uas_)
+	{
+		PluginBase::initialize(uas_);
+
+		adsb_nh.param<std::string>("frame_id", frame_id, "adsb_vehicle");
+
+		adsb_pub = adsb_nh.advertise<mavros_msgs::ADSBVehicle>("vehicle", 10);
+		adsb_sub = adsb_nh.subscribe("own", 10, &ADSBPlugin::adsb_cb, this);
+	}
+
+	Subscriptions get_subscriptions()
+	{
+		return {
+			       make_handler(&ADSBPlugin::handle_adsb)
+		};
+	}
+
+private:
+	ros::NodeHandle adsb_nh;
+
+	std::string frame_id;
+
+	ros::Publisher adsb_pub;
+	ros::Subscriber adsb_sub;
+
+	void handle_adsb(const mavlink::mavlink_message_t *msg, mavlink::common::msg::ADSB_VEHICLE &adsb)
+	{
+		auto adsb_msg = boost::make_shared<mavros_msgs::ADSBVehicle>();
+		const auto &adsb_callsign = adsb.callsign;
+
+		ros::Time tslc(adsb.tslc, 0);
+
+		adsb_msg->header.stamp = ros::Time::now();	//TODO: request add time_boot_ms to msg definition
+		adsb_msg->header.frame_id = frame_id;
+		adsb_msg->ICAO_address = adsb.ICAO_address;
+		adsb_msg->lla.latitude = adsb.lat / 1E7;
+		adsb_msg->lla.longitude = adsb.lon / 1E7;
+		adsb_msg->lla.altitude = adsb.altitude;	// in meters, TODO: #693
+		adsb_msg->altitude_type = adsb.altitude_type;
+		adsb_msg->heading = adsb.heading / 1E2 / 180.0f * M_PI - M_PI;
+		adsb_msg->hor_velocity = adsb.hor_velocity / 1E2;
+		adsb_msg->ver_velocity = adsb.ver_velocity / 1E2;	// up is positive
+		std::copy(adsb_callsign.cbegin(), adsb_callsign.cend(), adsb_msg->callsign.begin());
+		adsb_msg->emitter_type = adsb.emitter_type;
+		adsb_msg->tslc.data = tslc;
+		adsb_msg->data_status = adsb.flags;
+		adsb_msg->squawk = adsb.squawk;
+
+		// for info purposes
+		ROS_DEBUG_NAMED("adsb", "receiving data of altitude type: %s", utils::to_string(adsb.altitude_type).c_str());
+		ROS_DEBUG_NAMED("adsb", "receiving data of emitter type: %s", utils::to_string(adsb.emitter_type).c_str());
+		ROS_DEBUG_NAMED("adsb", "data status flags on receiving msg: %s", utils::to_string(adsb.flags).c_str());
+
+		adsb_pub.publish(adsb_msg);
+	}
+
+	void adsb_cb(const mavros_msgs::ADSBVehicle::ConstPtr &req)
+	{
```

Comment:
Use `ROS_DEBUG_NAMED("adsb", "ADSB: at type " << utils::to_string(e));` For enums! Use C++ streams for `std::string`.
Use `0x%02x` format for bit-filed.